Reorganize Python API for Garden Lattice book materials and add understanding book section (#223)
software-gardening/almanack: src/book/garden-lattice/understanding.md

# Understanding

```{figure} ../assets/garden-lattice-understanding-transfer.png
---
alt: Image showing two silhouettes with one showing transferrence of software idea
  to another.
---
_We transfer understanding of software projects in the form of documentation artifacts._
```

Shared understanding forms the heartbeat of any successful software project.
Peter Naur, in his work "Programming as Theory Building" (1985), emphasized that programming is fundamentally about building a shared theory of the project among all contributors {cite:p}`naur_programming_1985`.

> Understanding encompasses not only the technical aspects but also the collaborative and ethical dimensions that ensure a project's growth and sustainability.

Author: Thanks! More clear!

Just as a gardener must understand the needs of each plant to cultivate a thriving garden, contributors to a project must grasp its goals, structure, and community standards to foster a productive and harmonious environment.

> We present the following as beginnings towards shared understanding in a project.

Author: Thanks, a lot more descriptive!

## Common files for shared understanding

> In the realm of software development, certain files are commonly expected within a repository to ensure clarity, collaboration, and legal compliance.

Author: Thanks!

> These files serve as the foundational elements that guide contributors and users, much like a well-tended garden benefits from clear pathways and signs.

Author: Thank you!

> They also are correlated with increased rates of project success when present {cite:p}`coelho_why_2017`.

Author: Cheers, agreed - fascinating read!

> Historically the following files were written in plain-text without formatting.

Author: Thanks!

> More recently [markdown](https://en.wikipedia.org/wiki/Markdown) has become the preferred method of writing these files to help format the documents with styles and media.

Author: Thank you, more clear here!

GitHub, GitLab, and other software source control hosting platforms often automatically render markdown materials in HTML.

### README

The `README` file is the cornerstone of any repository, providing essential information about the project.

> Its presence signifies a well-documented and accessible project, inviting contributors and users alike to understand and engage with the work.

Author: Thanks!

A repository thrives when its purpose and usage are communicated through a `README` file, much like a garden benefits from a clear plan and understanding of its layout.

> "A good rule of thumb is to assume that the information contained within the README will be the only documentation your users read.
> For this reason, your README should include how to install and configure your software, where to find its full documentation, under what license it’s released, how to test it to ensure functionality, and acknowledgments.

> > Furthermore, you should include your quickstart guide (as introduced in Rule 3) in your README." {cite:p}`lee_ten_2018`

Author: Thanks, good point!

For further reading on `README` file practices, see the following resources:

- [The Turing Way: Landing Page - README File](https://book.the-turing-way.org/project-design/project-repo/project-repo-readme.html)
- [GitHub: About `README`s](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-readmes)
- [Makeareadme.com](https://www.makeareadme.com/)

### CONTRIBUTING

> A `CONTRIBUTING` file outlines the guidelines for contributing to the project.

Author: Thanks!

Its presence fosters a welcoming environment for new contributors, ensuring that they have the necessary information to participate effectively.

> In the same way that a garden flourishes when there's good understanding of roles and how to care for one another, a project grows stronger when contributors are guided by clear and inclusive instructions.

Author: Thanks, I'll work on this one. Generally I meant through collaboration without authority structure necessarily (not that this doesn't help, but that you can work together well with others without being told to do so by any particular person).

> All development procedures which are specific to the project should be outlined within the `CONTRIBUTING` file (for example, testing and linting expectations, release cadence, etc.).

Author: Thank you!

> Consider writing this documentation for each project from the standpoint of an outsider or beginner who may be trying to add a new task, understand the project, or complete an existing task {cite:p}`treude_towards_2024`.

Author: Thanks, yes, this still works I believe. The article talks about things through `newcomer` as a noun too, which can mean both of these things too (I like using both words to describe it to make it clear from multiple vantages).

For further reading on `CONTRIBUTING` file practices, see the following resources:

- [GitHub: Setting guidelines for repository contributors](https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/setting-guidelines-for-repository-contributors)
- [Mozilla: Wrangling Web Contributions: How to Build a CONTRIBUTING.md](https://mozillascience.github.io/working-open-workshop/contributing/)

### CODE_OF_CONDUCT

> The `CODE_OF_CONDUCT` (CoC) files set the standards for behavior within the project community.

Author: Thank you!

> Its presence is a commitment to maintaining a respectful and inclusive environment for all participants.

Reviewer: ```suggestion
Its presence signals a commitment to maintaining a respectful and inclusive environment for all participants.
```

Author: Nice touch here, thanks!

> It also may provide an outline for acceptable participation when it comes to conflicts of interest or similar.

Reviewer: ```suggestion
It also may provide an outline for acceptable participation when it comes to conflicts of interest.
```

Not sure what you mean by "similar" - consider either deleting or expanding with an example.

Author: I feel like there might be more here to this but I don't have strong evidence, I'll drop for now. Thanks!

> "CoCs are a place to express and define shared community values in OSS projects, an important governance document for moderation, and an indicator of project inclusivity and safety, especially for vulnerable contributors" {cite:p}`damian_codes_2024`

Reviewer: ```suggestion
A CoC is a foundational document that defines community values, guides governance and moderation, and signals inclusivity within the project, particularly for vulnerable contributors. {cite:p}`damian_codes_2024`
```

I recommend moving away from direct quotes unless absolutely necessary. Also, if you have direct quotes, they should be formatted differently

> "like this, I think" -Greg

Author: Thanks, we should move to your suggestion!

For further reading and examples of `CODE_OF_CONDUCT` files, see the following:

- [Contributor Covenant Code of Conduct](https://www.contributor-covenant.org/)
- [Example: Python Software Foundation Code of Conduct](https://policies.python.org/python.org/code-of-conduct/)
- [Example: Rust Code of Conduct](https://www.rust-lang.org/policies/code-of-conduct)

### LICENSE

A `LICENSE` file specifies the terms under which the project's code can be used, modified, and shared.

> "When you make a creative work (which includes code), the work is under exclusive copyright by default. Unless you include a license that specifies otherwise, nobody else can copy, distribute, or modify your work without being at risk of take-downs, shake-downs, or litigation. Once the work has other contributors (each a copyright holder), “nobody” starts including you." {cite:p}`choosealicense_nopermission`.
> The presence of a `LICENSE` file is crucial for legal clarity, encourages the responsible use or distribution of the project, and is considered an indicator of project maturity{cite:p}`deekshitha_rsmm_2024`.

> Similar to how a garden's health depends on understanding the rules of nature and respecting boundaries, a project's sustainability relies on clear licensing that protects both the creators and users, fostering a culture of trust and collaboration.

Reviewer: ```suggestion
Just as a garden's health depends on understanding natural laws and respecting boundaries, a project's sustainability hinges on clear licensing that safeguards and empowers both creators and users, cultivating a culture of trust and collaboration.
```

Author: Thank you!

For further reading and examples of `LICENSE` files, see the following:

- [OSF: Licensing](https://help.osf.io/article/148-licensing)
- [GitHub: Licensing a repository](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/licensing-a-repository)
- [Choosealicense.com](https://choosealicense.com/)
- [SPDX License List](https://spdx.org/licenses/)

## Project documentation

Common documentation files like `README`'s, `CONTRIBUTING` guides, and `LICENSE` files are only a start towards more specific project information.

> Comprehensive project documentation is akin to a detailed gardener's notebook for a well-maintained project, illustrating how the project may be used and guiding philosophy.

Reviewer: ```suggestion
Comprehensive project documentation is akin to a detailed gardener's notebook for a well-maintained project, illustrating how the project may be used and the guiding philosophy.
```

Author: Thanks!

> This includes in-depth explanations of the project's architecture, practical usage examples, API references, and development workflows.

Reviewer: prob should define API

Author: Thanks, will revise here based on this and the other comment.

Oftentimes this type of documentation is provided through a "documentation website" or "docsite" to facilitate a user experience that includes a search bar, multimedia, and other HTML-enabled features.

> Such documentation should strive to go above and beyond to ensure that both novice and seasoned contributors can grasp the project's complexities and contribute effectively.

Reviewer: I couldn't find the source that @MattsonCam posted in discord, but this sentence is in conflict with it. Essentially, good documentation (meaning actual `#`) should complement, but not be redundant, with readable code.

I wonder if there is a way to include this subtlety here, and add citation (Cameron, can you post the link here, if you remember what it was?)

Author: Thanks, yes, we should dig deeper here if there's a good citation to include! Generally I hope we can sift through to code comments or similar as metrics and guidance at some point.

Author: I spoke with @MattsonCam who mentioned that this is likely the link: https://accu.org/journals/overload/28/157/henney_2796/ . I'll work in this content somehow near here.

"We … find that tools appear to be more widely used when software developers provide deeper documentation, …" {cite:p}`afiaz_evaluation_2023`.
Just as a thriving garden benefits from meticulous care instructions and shared horticultural knowledge, a project flourishes when its documentation offers a clear and thorough guide to its inner workings, nurturing a collaborative and informed community.

Project documentation often exists within a dedicated `docs` directory where the materials may be created and versioned distinctly from other code.
Oftentimes this material will leverage a specific documentation tooling technology in alignment with the programming language(s) being used (for example, Python projects often leverage [Sphinx](https://www.sphinx-doc.org/en/master/)).

> These tools often increase the utility of the output by styling the material with pre-built HTML themes, automating application programming interface (API) documentation generation, and an ecosystem of plugins to help extend the capabilities of your documentation without writing new code.

Reviewer: ah, API defined here. prob should be switched to first reference

Author: Thanks will do!
